File: Recamans sequence - GFG/recamans-sequence.py
#User function Template for python3

class Solution:
    def recamanSequence(self, n):
        # code here
        dp=[0]*n 
        s=set()
        s.add(0)
        for i in range(1,n):
            val=dp[i-1]-i 
            #print(val)
            if val not in s and val>0:
                dp[i]=val
            else:
                dp[i]=dp[i-1]+i
            s.add(dp[i])
        return dp

#{ 
 # Driver Code Starts
#Initial Template for Python 3

if __name__ == '__main__':
    t = int(input())
    for _ in range(t):
        n = int(input())
        
        ob = Solution()
        ans = ob.recamanSequence(n)
        for i in range(n):
            print(ans[i],end=" ")
        print()
# } Driver Code Ends

File: Recamans sequence - GFG/recamans-sequence.cpp
//{ Driver Code Starts
// Initial Template for C++

#include <bits/stdc++.h>
using namespace std;

// } Driver Code Ends
// User function Template for C++

class Solution{
public:

    int sol(vector<int>&dp,int n){
        if(n==0) return dp[0];
        // if(n==1) return dp[1];
        if(dp[n]!=-1) return dp[n];
        int x=sol(dp,n-1)-n;
        int c=0;
        for(int i=0;i<dp.size();i++){
            if(dp[i]==x)c=1;
        }
        if(x>0&&c==0) dp[n]=x;
        else dp[n]=sol(dp,n-1)+n;
    }
    vector<int> recamanSequence(int n){
        // code here
        vector<int>dp(n+1,-1);
        dp[0]=0;
        dp[1]=1;
        sol(dp,n);
        return dp;
    }
};

//{ Driver Code Starts.

int main(){
    int t;
    cin>>t;
    while(t--){
        int n;
        cin>>n;
        
        Solution ob;
        vector<int> ans = ob.recamanSequence(n);
        for(int i = 0;i < n;i++)
            cout<<ans[i]<<" ";
        cout<<"\n";
    }
    return 0;
}
// } Driver Code Ends